### DISCUSSION TITLE
Update Dont Work For AI THinker board

### DISCUSSION THREAD

**Discussion by afshinjian** [Hieromon/AutoConnect#414] [Q&A] [answered]
Hi Hieromon
Thank you for doing a very good project.
The update section does not work for the AI Thinker ESP32 cam model.
My code is too long 

[rest2-OTA-autoconnect.zip](https://github.com/Hieromon/AutoConnect/files/7316197/rest2-OTA-autoconnect.zip)

**Comment by Hieromon** [comment]
See [here](https://hieromon.github.io/AutoConnect/adothers.html#debug-print) also [FAQ](https://hieromon.github.io/AutoConnect/faq.html#3-turn-on-the-debug-log-options), and get the **AC_DEBUG** log. I don't review the whole of your code, but I can parse the AC_DEBUG log.
First, verify with you that your sketch meets the requirements for OTA.
In addition, [this area](https://arduino.stackexchange.com/questions/75198/why-doesnt-ota-work-with-the-ai-thinker-esp32-cam-board) or [this REPO](https://github.com/wjsanek/wjsanek) may help you.

By the way, the Web Server required to satisfy all the functions of ESP32-CAM is different from the ESP32 WebServer library. AutoConnect does not officially support it. It may be applicable if you hack it, but it's not included in my plan.

**Comment by afshinjian** [comment]
Thank you very much for your accurate response and great tips.
This guide was exactly right :
As stated in the comments by @Codebreaker007 using the ESP32 Wrover Module works and if you put the pin definitions inline in the app there isn't a problem
and then insert the sketch:
#define CAMERA_MODEL_AI_THINKER

#if defined(CAMERA_MODEL_AI_THINKER)
  #define PWDN_GPIO_NUM     32
  #define RESET_GPIO_NUM    -1
  #define XCLK_GPIO_NUM      0
  #define SIOD_GPIO_NUM     26
  #define SIOC_GPIO_NUM     27
  
  #define Y9_GPIO_NUM       35
  #define Y8_GPIO_NUM       34
  #define Y7_GPIO_NUM       39
  #define Y6_GPIO_NUM       36
  #define Y5_GPIO_NUM       21
  #define Y4_GPIO_NUM       19
  #define Y3_GPIO_NUM       18
  #define Y2_GPIO_NUM        5
  #define VSYNC_GPIO_NUM    25
  #define HREF_GPIO_NUM     23
  #define PCLK_GPIO_NUM     22
#else
  #error "Camera model not selected"
#endif

Two important points :
1st is The settings should look like this

UNDER ">TOOLS>"
THIS SKETCH USED WITH THESE SETTINGS:
Board: ESP32 Wrover Module
Upload Speed: 115200
Flash Frequency: 80MHz
Flash Mode: QIO
Partition Scheme: Minimal SPIFFS (1.9MB APP with OTA/190KB SPIFFS)
Core 
Debug Level: None
Port: "Based on the device"

and 2nd is :
If you use a HTTP client or server in the program, you must connect the GPIO0 to GND before starting the update, and after the update, disconnect the GPIO0  from the GND and then restart.
Thank you for your kindness, it was a great help Hieromon

**Comment by Hieromon** [comment]
@afshinjian I'm glad you have taken a step toward the solution. I recognized that the "**The update section doesn't work**" you said was **AutoConnect's OTA**. Because [AutoConnect OTA](https://hieromon.github.io/AutoConnect/otabrowser.html#updates-with-the-web-browserupdated-wv115) will show "**Update**" in the menu. On this premise, I have a few questions for you.

1. Partition scheme
Refs: [**AutoConnect FAQ**](https://hieromon.github.io/AutoConnect/faq.html#how-much-memory-does-autoconnect-consume)
You were told that choosing a partition schema was a hint, so I don' know you were choosing the default.csv from the beginning.     But that shouldn't matter directly related to OTA. Because if the compiled binary size exceeds the default size, a link error will occur. If it is not an error, the space required for OTA is reserved. So you can't get a sketch beyond the partition. But you are saying "update section doesn't work". This means that you already have a compiled sketch binary of a size that can accommodate the default partition.    
The definitions below are the `Default` and `Minimal SPIFFS` schemas. The `app0` is the flash area for the sketch binary. This area will be boot. OTA binary will be staged to `app1`. And after reset, it will be copied to `app0`. This is how ESP8266 and ESP32 OTA work. The `app0` and `app1` have the same value in both of the two partition definitions. That is, if you are using the Arduino IDE to create a sketch binary that fits in `app0`, there is already an OTA area (`app1`) to reserved it.

- Default partition scheme
    https://github.com/espressif/arduino-esp32/blob/master/tools/partitions/default.csv
    ```
    # Name,   Type, SubType, Offset,  Size, Flags
    nvs,      data, nvs,     0x9000,  0x5000,
    otadata,  data, ota,     0xe000,  0x2000,
    app0,     app,  ota_0,   0x10000, 0x140000,
    app1,     app,  ota_1,   0x150000,0x140000,
    spiffs,   data, spiffs,  0x290000,0x170000,
    ```

- Minimal SPIFFS
    https://github.com/espressif/arduino-esp32/blob/master/tools/partitions/min_spiffs.csv
    ```
    # Name,   Type, SubType, Offset,  Size, Flags
    nvs,      data, nvs,     0x9000,  0x5000,
    otadata,  data, ota,     0xe000,  0x2000,
    app0,     app,  ota_0,   0x10000, 0x1E0000,
    app1,     app,  ota_1,   0x1F0000,0x1E0000,
    spiffs,   data, spiffs,  0x3D0000,0x30000,
    ```
    
    Therefore, the problem you encounter is not directly related to the partition schema. am I wrong?

2. GPIO0 to flash the sketch binary
Controlling the `GPIO0` pin is required to switch to ESP32 flash mode. Also, we need to trigger the `EN` (reset) pin to boot automatically after flushing. A typical ESP32 programmer with USB serial, such as the ESP-Prog, controls these two signals with RS-232C DTR / RTS signals. This trick allows flash from a USB host and post-flash reset and serial communication. The ESP32-DevKit has this circuit recommended by [Espressif](https://docs.espressif.com/projects/espressif-esp-iot-solution/en/latest/hw-reference/ESP-Prog_guide.html#automatic-downloading-function), but the ESP32-CAM standalone module lacks it.

- The flash programing Circuit
    https://docs.espressif.com/projects/espressif-esp-iot-solution/en/latest/hw-reference/ESP-Prog_guide.html#automatic-downloading-function
    <img width="480px" src="https://docs.espressif.com/projects/espressif-esp-iot-solution/en/latest/_images/Auto_program.png">

 - Ai-Thinker ESP32-CAM Schematic
    https://docs.ai-thinker.com/_media/esp32/docs/esp32_cam_sch.pdf
    ![esp32_cam_sch](https://user-images.githubusercontent.com/12591771/136776019-bc993a51-713b-44c1-977e-453acc0eece5.png)

    I know there is the [ESP32-CAM AI-Thinker MB Programmer](https://makeradvisor.com/tools/esp32-cam-mb-usb-programmer/) that can supplement the flash programming circuit of the ESP32-CAM, but I don't know if you are using it.
    Refs: [Upload Code to ESP32-CAM AI-Thinker using ESP32-CAM-MB USB Programmer](https://randomnerdtutorials.com/upload-code-esp32-cam-mb-usb/)
    However, the signal control required for this flash programming has nothing to do with OTA. It's only needed for serial programming. So if your "The update section does not work" is OTA, then the `GPIO0` connection with `GND` has nothing to do with the problem you encountered.

So, Could you describe your problem clearly and specifically? Does your "**update**" mean the serial flash programming? (that's Arduino IDE menu Sketch > **Upload**)

**Comment by Hieromon** [comment, marked as answer]
@afshinjian I felt your question wasn't right and reluctantly skimmed your code. As a result, there are some unsuitable points in your code as followings:

1. The function definition is incorrect
    The `checkconnection` function declares with the `String` type, but it never returns the String value. Its definition must be the `void` type. Same as the `testSitu`. 
2. Attempts to perform a long delay that breaks the `handleClent` loop
- In `loop` function
    ```
    Portal.handleClient();
    digitalWrite(16, LOW);
    delay (1500);    // <-- Disturb the handleClient loop for WebServer behavior
    digitalWrite(16, HIGH);
    ```
    ```
    while (count > 0) {
      sendPhoto();
      Serial.println("Sending IMAGE to SERVER");
      delay (20000);  // <-- 20 seconds! During that period, the ESP32 Web Server (that is, it is an AutoConnect menu behavior) cannot respond to requests from clients).
      count = count - 1 ;
    }
    ```
    And many many `delay`...
- In `sendPhoto`, `sendCapturedImage2LineNotify` function
```
    int waitTime = 10000;   // timeout 10 seconds
    long startTime = millis();
    boolean state = false;

    while ((startTime + waitTime) > millis())
    {
      Serial.print(".");
      delay(100);
      while (client_tcp.available())
      {
        char c = client_tcp.read();
        if (c == '\n')
        {
          if (getAll.length() == 0) state = true;
          getAll = "";
        }
        else if (c != '\r')
          getAll += String(c);
        if (state == true) getBody += String(c);
        startTime = millis();
      }
      if (getBody.length() > 0) break;
    }
```

It combines without thinking much about that the captured image throwing logic which only **acts as a client model**, with the handler process for the WebServer which should **act as the server model**. This code doesn't work even if you exclude AutoConnect and replace it with the WebServer library.
I found out that you are partially diverting the code posted on a particular site and splicing the code. https://randomnerdtutorials.com/esp32-cam-post-image-photo-server/
The result is the above nonconformity If you're pasting code from here and there, you need to understand the programming model for the Web Server and make the appropriate changes. Your code is inconsistent. If you post an issue, make sure you understand how the sketch works and narrow down the issues with AutoConnect's behavior.
Ref: [Reports the issue to AutoConnect](https://hieromon.github.io/AutoConnect/faq.html#4-reports-the-issue-to-autoconnect-github-repository)

This topic is not suitable as an issue, so move on to the Discussions Q&A category.
